src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+import {
+  AddressType,
+  AddressValidationResult,
+  CourtAddress,
+  DisplayAddress,
+  DisplayCourtAddresses
+} from '../../../types/CourtAddress';
+import {CourtAddressPageData} from '../../../types/CourtAddressPageData';
+import {SelectItem} from '../../../types/CourtPageData';
+
+@autobind
+export class AddressController {
+
+  getAddressTypesError = 'A problem occurred when retrieving the court address types.';
+  getAddressesError = 'A problem occurred when retrieving the court addresses.';
+  updateAddressError = 'A problem occurred when saving the court addresses.';
+  multipleVisitAddressError = 'Only one visit address is permitted.';
+  typeRequiredError = 'Address Type is required.';
+  addressRequiredError = 'Address is required.';
+  townRequiredError = 'Town is required.';
+  invalidPostcodeError = 'Postcode is invalid.';
+  postcodeMissingError = 'Postcode is required.';
+  postcodeNotFoundError = 'Postcode entered could not be found.';
+  writeToUsAddressType = 'Write to us';
+  visitOrContactUsAddressType = 'Visit or contact us';
+  primaryAddressPrefix = 'Primary Address: ';
+  secondaryAddressPrefix = 'Secondary Address: ';
+
+  public async get(req: AuthedRequest, res: Response): Promise<void> {
+    await this.render(req, res);
+  }
+
+  public async put(req: AuthedRequest, res: Response): Promise<void> {
+    const addresses: DisplayCourtAddresses = {
+      primary: req.body.primary,
+      secondary: req.body.secondary,
+    };
+    const writeToUsTypeId = req.body.writeToUsTypeId;
+
+    // Validate token
+    if(!CSRF.verify(req.body._csrf)) {
+      await this.render(req, res, false, addresses, [this.updateAddressError]);
+      return;
+    }
+
+    // Validate addresses
+    const primaryValidationResult = this.validateAddress(addresses.primary, true);
+    const secondaryValidationResult = this.validateAddress(addresses.secondary, false);
+    const addressTypeErrors  = this.validateAddressTypes(addresses.primary, addresses.secondary, writeToUsTypeId);
+    const addressesValid = primaryValidationResult.addressValid && primaryValidationResult.postcodeValid &&
+      secondaryValidationResult.addressValid && secondaryValidationResult.postcodeValid && addressTypeErrors.length === 0;
+
+    if (!addressesValid) {
+      const allErrors = primaryValidationResult.errors.concat(secondaryValidationResult.errors).concat(addressTypeErrors);
+      await this.render(req, res, false, addresses, allErrors,
+        !primaryValidationResult.postcodeValid, !secondaryValidationResult.postcodeValid);
+      return;
+    }
+
+    // Post addresses to API if valid
+    await req.scope.cradle.api.updateCourtAddresses(req.params.slug, this.convertToApiType(addresses))
+      .then(async (addressList: CourtAddress[]) => await this.render(req, res, true, this.convertToDisplayAddresses(addressList)) )
+      .catch(async (reason: AxiosError) => {
+        if (reason.response.status === 400) {
```

Comment:
Just double checking: Is there a possibility of a duplicate? I.e a 409 response 

If I added in: 

Andrew Street 20, Scotland has nice hills town, AB37

And then the same again, would that be allowed? I think you may be checking for duplicated addresses. 

Another question on the same train of thought would be: should that check also be in the API if other teams may reuse the endpoint? And if so, we should or could cater for the response status being returned here

Reply:
This other one is up to you entirely (don't worry about it if you think its time consuming, or if it isn't warrented): https://www.npmjs.com/package/http-status-codes , if we want to start using stuff like: `StatusCodes.NOT_FOUND`

Reply:
We don't have any 409 or similar responses in the API currently. I'm pretty sure it was confirmed we don't need to do duplicate checking.